hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/ObserverReadProxyProvider.java

```diff
@@ -648,6 +645,7 @@ public synchronized void close() throws IOException {
       }
     }
     failoverProxy.close();
+    nnProbingThreadPool.shutdown();
```

Comment:
is close ever called?

Reply:
According to this post, it would be called when the garbage collector is reclaiming that object. When that would happen is out of control from programmers. 

https://stackoverflow.com/questions/18670658/why-close-cannot-be-called-on-closeable-classes-automatically

Reply:
thats not going to solve the issue right? The issue is the job is hanging until this is closed, right? so waiting for GC to call close isn't sufficient right? We expect/want this to close and finish immediately.

Reply:
We switched to use BlockingThreadPoolExecutorService, which set `allowCoreThreadTimeout` to true. When the main thread exits, there won't be new tasks submitted to the executorService. Then all threads, including the core threads will be idle and they will all be shut down, due to the allowCoreThreadTimeout property to be true. 

After all threads from the executorService are shutdown, and there is no active thread, JVM will be shutdown.

Reply:
The issue is previously we did not set _allowCoreThreadTimeout_ to true. So, the core thread will be kept running, though there is no new tasks submitted. As long as there is a thread running, JVM won't be shut down (unless we force it with an explicit `System.exit()`).

Reply:
so whats the expected behavior? We finish, then it'll idle for how many seconds (blocking)?
also its not obvious to me that the main thread timing out or the GC calling close is the thing that we will be blocked on

Reply:
The expected behavior is we will wait for the main thread to exit and the threads in the executorService pool to shut down (after idle for 10 seconds). Then, the JVM process will be shut down. In the worst case, we may take 10 more seconds to shut down the JVM process.